### PR TITLE
Implement `Span::sub_span` and `Span::lines_span`.

### PR DESCRIPTION
**Pull request by tangmi**
Prototype implementation for #455

### CODE REVIEW

**Comment by CAD97** [comment]
This looks reasonable to me at a glance. @dragostis ?

**Comment by tangmi** [comment]
Some concerns I have (that may not actually be real issues):
- Is `Span` intended to be consumable by users in this way? My assumption is that pest expects/allows users to create `Span`, given that `Span::new` is exposed (although I'm proposing this PR because `Span::new` is difficult to use with a parsed document since `Span::input` is not exposed).
- I didn't put much thought into the naming of new members
  - `Span::sub_span` would be called `Span::get` to parallel `str::get` (there is no "substring" method in `std`, just range indexing)
  - `LinesSpan` was just named as such to avoid changing the name of `Lines`.

**Comment by Alextopher** [comment]
I am also interested in `sub_span`. I am also happy to provide some kind of "help wanted" to see it through

**Comment by Alextopher** [comment]
> Is Span intended to be consumable by users in this way?
There are similar requests for additional flexibility ( such as #469 ). There are other types that provide flexibility such as [ErrorVariant::CustomError](https://docs.rs/pest/2.2.1/pest/error/enum.ErrorVariant.html#variant.CustomError). Subspans could be similarly useful.

> `Span::sub_span` would be called `Span::get` to parallel `str::get`
I prefer `Span::get` to match other subslicing behaviors

`LinesSpan` seems fine to me, it should only ever be used as an iterator.

@tomtau is this PR blocked waiting for better documentation?

**Comment by tomtau** [comment]
@Alextopher unless @tangmi plans to revive this PR soon, yes, I think two main items to fix will be to adjust the docs a little bit (to provide a full example usage) and to resolve a minor conflict with the latest target branch.
@Alextopher feel free to open a new PR based on this one.

File: pest/src/span.rs [hunk 1]
```diff
@@ -62,6 +62,32 @@ impl<'i> Span<'i> {
...
+            std::ops::Bound::Excluded(&offset) => offset + 1,
+            std::ops::Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            std::ops::Bound::Included(&offset) => Some(offset + 1),
+            std::ops::Bound::Excluded(&offset) => Some(offset),
+            std::ops::Bound::Unbounded => None,
+        };
+        let s = self.as_str();
+        if s.get(start..end.unwrap_or(s.len())).is_some() {
```
Comment (tomtau) [inline pest/src/span.rs:80]: @Alextopher one more item to fix: the merge conflict seems to be about `core` vs `std`. pest now supports `no_std`, so these `ops` may either need to be adjusted for `no_std` or sub_span should be feature-guarded for `std`

File: pest/src/span.rs [hunk 2]
```diff
@@ -239,30 +275,82 @@ impl<'i> Hash for Span<'i> {
...
         if pos.at_end() {
             return None;
         }
-        let line = pos.line_of();
+        let line_start = self.pos;
         self.pos = pos.find_line_end();
-        Some(line)
+        Some(unsafe {
+            Span::new_unchecked(self.span.input, line_start, self.pos.min(self.span.end))
+        })
```
Comment (tomtau) [inline pest/src/span.rs:310]: @Alextopher could this be done without `unsafe`? if not, it should have a `Safety` comment explaining why it is safe or when things could go wrong

**Comment by Alextopher** [comment]
Sounds good. I started by creating a new PR with just `Span::get` and will start looking into `LinesSpan` now.

**Comment by tomtau** [comment]
closing in favour of https://github.com/pest-parser/pest/pull/681

**Pull request closed without merging**